pkg/operator/controllers/cloudproviderconfig/cloudproviderconfig_controller.go

```diff
@@ -0,0 +1,236 @@
+			Log:    Log,
+			Client: client,
+			Name:   ControllerName,
+		},
+	}
+}
+
+// Reconcile makes sure that the cloud-provider-config is healthy
+func (r *CloudProviderConfigReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	r.Log.Debug("reconcile ConfigMap openshift-config/cloud-provider-config")
+
+	instance, err := r.GetCluster(ctx)
+	if err != nil {
+		r.Log.Info(err)
+		return reconcile.Result{}, err
+	}
+
+	if !instance.Spec.OperatorFlags.GetSimpleBoolean(controllerEnabled) {
+		r.Log.Info("controller is disabled")
+		return reconcile.Result{}, nil
+	}
+
+	r.Log.Debug("running")
+	return reconcile.Result{}, r.updateCloudProviderConfig(ctx)
+}
+
+// SetupWithManager setup our manager
+func (r *CloudProviderConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.Log.Info("starting cloud-provider-config controller")
+
+	aroClusterPredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return o.GetName() == arov1alpha1.SingletonClusterName
+	})
+
+	cloudProviderConfigPredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return o.GetName() == cloudProviderConfigName.Name && o.GetNamespace() == cloudProviderConfigName.Namespace
+	})
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&arov1alpha1.Cluster{}, builder.WithPredicates(aroClusterPredicate)).
+		Watches(
+			&source.Kind{Type: &corev1.ConfigMap{}},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(cloudProviderConfigPredicate),
+		).
+		Named(ControllerName).
+		Complete(r)
+}
+
+func (r *CloudProviderConfigReconciler) updateCloudProviderConfig(ctx context.Context) error {
+	r.Log.Debug("checking openshift-config/cloud-provider-config")
+
+	cm := &corev1.ConfigMap{}
+	err := r.Client.Get(ctx, cloudProviderConfigName, cm)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			r.Log.Debug("the ConfigMap cloud-provider-config was not found in the openshift-config namespace")
+		}
+		return err
+	}
+
+	jsonConfig, ok := cm.Data["config"]
+	if !ok {
+		return fmt.Errorf("field config in ConfigMap openshift-config/cloud-provider-config is missing")
+	}
+
+	changed := false
+	var cpc azCloudProviderConfig
+	err = json.Unmarshal([]byte(jsonConfig), &cpc)
+	if err != nil {
+		return err
+	}
+
+	if cpc.DisableOutboundSNAT != nil && !*cpc.DisableOutboundSNAT {
+		r.Log.Info("updating openshift-config/cloud-provider-config disableOutboundSNAT from false to true")
+		*cpc.DisableOutboundSNAT = true
+		changed = true
+	} else if cpc.DisableOutboundSNAT == nil {
+		r.Log.Info("updating openshift-config/cloud-provider-config disableOutboundSNAT from nil to true")
+		truePointer := true
+		cpc.DisableOutboundSNAT = &truePointer
+		changed = true
+	} else {
+		r.Log.Debug("openshift-config/cloud-provider-config disableOutboundSNAT is set to true no changes needed")
+	}
+
+	jsonStringByte, err := json.Marshal(cpc)
+	if err != nil {
+		return err
+	}
+
+	cm.Data["config"] = string(jsonStringByte)
```

Comment:
This data was read as bytes, but is being written as a string, is that ok?  I did test this change and the load balancer was updated as expected and everything worked great, I'm just wanting to double check this with you as I noticed the config map went from this
```
data:
  config: "{\n\t\"cloud\": \"AzurePublicCloud\",\n\t\"tenantId\": \"64dc69e4-d083-49fc-9569-ebece1dd1408\",\n\t\"aadClientId\":
    \"\",\n\t\"aadClientSecret\": \"\",\n\t\"aadClientCertPath\": \"\",\n\t\"aadClientCertPassword\":
    \"\",\n\t\"useManagedIdentityExtension\": false,\n\t\"userAssignedIdentityID\":
    \"\",\n\t\"subscriptionId\": \"fe16a035-e540-4ab7-80d9-373fa9a3d6ae\",\n\t\"resourceGroup\":
    \"aro-reo2fmqn\",\n\t\"location\": \"eastus\",\n\t\"vnetName\": \"aro-vnet\",\n\t\"vnetResourceGroup\":
    \"tschneid-02\",\n\t\"subnetName\": \"worker-subnet\",\n\t\"securityGroupName\":
    \"tschneid-npmdm-nsg\",\n\t\"routeTableName\": \"tschneid-npmdm-node-routetable\",\n\t\"primaryAvailabilitySetName\":
    \"\",\n\t\"vmType\": \"\",\n\t\"primaryScaleSetName\": \"\",\n\t\"cloudProviderBackoff\":
    true,\n\t\"cloudProviderBackoffRetries\": 0,\n\t\"cloudProviderBackoffExponent\":
    0,\n\t\"cloudProviderBackoffDuration\": 6,\n\t\"cloudProviderBackoffJitter\":
    0,\n\t\"cloudProviderRateLimit\": false,\n\t\"cloudProviderRateLimitQPS\": 0,\n\t\"cloudProviderRateLimitBucket\":
    0,\n\t\"cloudProviderRateLimitQPSWrite\": 0,\n\t\"cloudProviderRateLimitBucketWrite\":
    0,\n\t\"useInstanceMetadata\": true,\n\t\"loadBalancerSku\": \"standard\",\n\t\"excludeMasterFromStandardLB\":
    false,\n\t\"disableOutboundSNAT\": null,\n\t\"maximumLoadBalancerRuleCount\":
    0\n}\n"
```
to this
```
data:
  config: '{"cloud":"AzurePublicCloud","tenantId":"64dc69e4-d083-49fc-9569-ebece1dd1408","aadClientId":"","aadClientSecret":"","aadClientCertPath":"","aadClientCertPassw
ord":"","useManagedIdentityExtension":false,"userAssignedIdentityID":"","subscriptionId":"fe16a035-e540-4ab7-80d9-373fa9a3d6ae","resourceGroup":"aro-reo2fmqn","location"
:"eastus","vnetName":"aro-vnet","vnetResourceGroup":"tschneid-02","subnetName":"worker-subnet","securityGroupName":"tschneid-npmdm-nsg","routeTableName":"tschneid-npmdm-
node-routetable","primaryAvailabilitySetName":"","vmType":"","primaryScaleSetName":"","cloudProviderBackoff":true,"cloudProviderBackoffRetries":0,"cloudProviderBackoffEx
ponent":0,"cloudProviderBackoffDuration":6,"cloudProviderBackoffJitter":0,"cloudProviderRateLimit":false,"cloudProviderRateLimitQPS":0,"cloudProviderRateLimitBucket":0,"
cloudProviderRateLimitQPSWrite":0,"cloudProviderRateLimitBucketWrite":0,"useInstanceMetadata":true,"loadBalancerSku":"standard","excludeMasterFromStandardLB":false,"disa
bleOutboundSNAT":true,"maximumLoadBalancerRuleCount":0}'
```

Reply:
we might be able to test this out functionally in an E2E? Attempt to modify the CM and then see that it gets reverted? Granted, idk what kind of timeframe this would take :/

Reply:
I tested this scenario with the functional testing.